### PR TITLE
Implement Ellipse type

### PR DESCRIPTION
**Pull request by mworleybr**

### CODE REVIEW

File: src/ellipse.rs [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+        let r_squared = self.radius_squared();
+        let sum = offset_squared.dx / r_squared.width() + offset_squared.dy / r_squared.height();
+        sum < T::one() || float_cmp::approx_eq!(T, sum, T::one())
+    }
+
+    pub fn bounding_rect(&self) -> Rect<T> {
+        let offset = self.radius.to_vector();
+        let top_left = self.center - offset;
+        let bottom_right = self.center + offset;
+        Rect::from_points(top_left, bottom_right)
```
Comment (francesca64) [inline src/ellipse.rs]: Could this just be `Rect::from_center(self.center, self.radius)`?

File: src/ellipse.rs [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+
+    pub fn map_radius(self, f: impl FnOnce(Size<T>) -> Size<T>) -> Self {
+        self.map(move |center, radius| (center, f(radius)))
+    }
+
+    pub fn cast<U: en::Num>(self) -> Ellipse<U> {
+        self.map(move |center, radius| {
+            (
+                center.cast(),
+                Size::new(en::cast(radius.width()), en::cast(radius.height())),
```
Comment (francesca64) [inline src/ellipse.rs]:
```suggestion
                radius.cast(),
```

File: src/ellipse.rs [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+        self.radius
+    }
+
+    pub fn radius_squared(&self) -> Size<T> {
+        self.radius * self.radius.to_vector()
+    }
+
+    pub fn contains(&self, point: Point<T>) -> bool
+    where
+        T: float_cmp::ApproxEq,
```
Comment (francesca64) [inline src/ellipse.rs]: `en::Num` and `en::Float` are intended to be the alpha and omega of numeric traits, so it's generally verboten in gee to constrain on anything else. If this crate is essential, you'd need to incorporate it into the `en::Float` trait.

**Review by francesca64 — CHANGES_REQUESTED** [review]
Thanks for adding tests too!

File: src/ellipse.rs (hunk 1 above)
Comment (mworleybr) [reply src/ellipse.rs]: `self.radius * 2`, but you're right.  Simplified!

File: src/ellipse.rs (hunk 3 above)
Comment (mworleybr) [reply src/ellipse.rs]: Nahh, wasn't essential, just didn't realize there was an epsilon() in `en::Float` that I could use.

**Comment by francesca64** [comment]
`Ellipse` should impl `From<Circle>`, and also have the associated fn `from_circle` along with `Circle` having `to_ellipse` (explicit conversions like that are often more convenient for satisfying the typechecker i.e. in the middle of a method chain)

After that, we can use `to_ellipse` in `bounding_rect`, `arc_points`, and `circle_points` to avoid duplication of math that's very nearly the same.

Part of me thinks `Circle` should just wrap `Ellipse`, but most of these functions are so trivial that the boilerplate of forwarding all of them could outweigh the benefits. Using `to_ellipse` to opportunistically share math seems like the happy medium.

File: src/ellipse.rs [hunk 4]
```diff
@@ -160,6 +164,12 @@ impl<T: en::Num> SubAssign<Vector<T>> for Ellipse<T> {
     }
 }
 
+impl<T: en::Num> From<Circle<T>> for Ellipse<T> {
+    fn from(circle: Circle<T>) -> Ellipse<T> {
+        Ellipse::new(circle.center(), Size::square(circle.radius()))
```
Comment (francesca64) [inline src/ellipse.rs]:
```suggestion
        Self::new(circle.center(), Size::square(circle.radius()))
```

**Review by francesca64 — CHANGES_REQUESTED** [review]

**Review by francesca64 — APPROVED** [review]
Thanks!

**Pull request merged**